### PR TITLE
feat(sft): add WAN2.1 UCF-101 full-transformer SFT

### PR DESCRIPTION
**Pull request by leviking98z-rgb**
## Summary

- Add target-video supervised fine-tuning for WAN2.1: bounded video decode -> WAN VAE clean latent -> flow-matching SFT.
- Add a reproducible **UCF-101 full-transformer SFT** recipe and deterministic manifest builder.
- Reuse one bounded PyAV/tensor video loader across SFT, WAN V2V, and Qwen3-Omni while keeping the SFT builder as a thin manifest-to-Part adapter.
- Make PyAV available to the core Linux install for worker-side target-video decoding.

## Related Issue

N/A

## UCF-101 Full-Transformer SFT

Model and dataset:

```text
WAN2.1-T2V-1.3B
UCF-101 18-class sports/action subset
Full transformer finetuning
VAE and text encoder frozen
```

Training setup:

```text
2,367 training videos / 128 held-out validation videos
17 frames, 256x448
1 node x 8 NVIDIA H20 96 GB GPUs
Global batch size 8
Learning rate 2e-6
300 optimizer steps
Checkpoints at steps 100, 200, and 300
```

Results:

```text
Eval loss at initialization: 0.21981
Best eval loss, step 260: 0.16230  (-26.2%)
Final eval loss, step 300: 0.16264
Final train loss: 0.16008
Final grad norm: 0.30942
```

### Training curves

The screenshot comes from the completed W&B run. `eval/loss` drops sharply in the first 20 steps and continues improving to its best value at step 260. Raw per-step `train/loss` is noisy because each diffusion batch samples a different timestep; `train/loss_ema_50` shows the underlying downward trend.

WAN2.1 UCF-101 Full SFT W&B training curves

W&B run: [`leviking98z-zhejiang-university/unirl-pr320-wan21-video-sft/jgifozoc`](https://wandb.ai/leviking98z-zhejiang-university/unirl-pr320-wan21-video-sft/runs/jgifozoc)

### Qualitative comparison

The matched-prompt/matched-seed panels below compare the untouched WAN2.1 base model with the step-300 full checkpoint at START / MIDDLE / END. They demonstrate UCF-101 distribution adaptation rather than reconstruction of a source clip.

**Archery**

WAN2.1 Base vs UCF-101 Full SFT: archery

**Surfing**

WAN2.1 Base vs UCF-101 Full SFT: surfing

**Basketball dunk**

WAN2.1 Base vs UCF-101 Full SFT: basketball dunk

**Horse riding**

WAN2.1 Base vs UCF-101 Full SFT: horse riding

**Diving**

WAN2.1 Base vs UCF-101 Full SFT: diving

## Reproduce

Prepare manifests:

```bash
hf download quchenyuan/UCF101-ZIP UCF-101.zip \
  --repo-type dataset --local-dir /path/to/ucf101
unzip /path/to/ucf101/UCF-101.zip -d /path/to/ucf101
python -m unirl.utils.prepare_sft_t2v_ucf101 \
  --data-root /path/to/ucf101/UCF-101 \
  --out-dir data/sft_ucf101_sports
```

Train:

```bash
ENTRY=train_sft \
SFT_DATA=data/sft_ucf101_sports/train.jsonl \
SFT_EVAL_DATA=data/sft_ucf101_sports/val.jsonl \
OUTPUT_DIR=outputs/wan21_t2v_ucf101_full \
bash examples/run_experiment_single_node.sh sft/wan21_t2v_ucf101_full
```

## Test Plan

- Full 300-step training completed without OOM, NaN, or Inf on 1 node x 8 NVIDIA H20 96 GB GPUs.
- Full checkpoints were successfully written at steps 100, 200, and 300.
- `SKIP=no-commit-to-branch pre-commit run --all-files --show-diff-on-failure` passes locally.
- The UCF-101 manifest builder was checked with a deterministic 18-class fixture.
- Focused smoke tests cover `.pt`/`.npy`/`.npz` and MP4 loading, full-clip bounded sampling, pipeline-owned WAN target encoder construction, and WAN2.2 preprocessing parity.
- PR images are hosted outside the PR diff; this PR adds no `assets/` files.

## Follow-up TODO

- [ ] Handle FSDP2 versions that do not expose the private `FSDPModule._orig_cls_mro_index` attribute in a separate compatibility PR. The previous fallback was removed from this feature PR because it changes the shared checkpoint-loading path; that follow-up should include a focused regression test and an explicit supported PyTorch version range.

## Compatibility / Risk

- The new WAN target-video path requires PyAV for worker-side decoding.
- Full-transformer SFT is GPU-memory intensive; the demonstrated setup uses 8 H20 96 GB GPUs with FSDP.
- Existing non-SFT WAN recipes are unchanged.

## Reviewer Notes

- This PR intentionally contains only the UCF-101 full-transformer SFT baseline. Self-Forcing/DMD work is isolated in a separate branch and PR.

## Checklist

- [x] I reviewed the changed code and removed unrelated/generated artifacts.
- [x] I updated tests, docs, and configs where needed, or explained why not.

### CODE REVIEW

**Review by haonan3 — APPROVED** [review]
LGTM
